docs/user-guide/05-compiling.md

```diff
@@ -0,0 +1,493 @@
+For example, if `MyShaders.slang` contained:
+
+```hlsl
+[shader("compute")]
+void myComputeMain(...) { ... }
+```
+
+then the Slang system will automatically detect and validate this entry point as part of a `loadModule("MyShaders")` call.
+
+After a module has been loaded, the application can look up entry points in that module using `IModule::findEntryPointByName()`:
+
+```c++
+SlangComPtr<IEntryPoint> computeEntryPoint;
+module->findEntryPointByName("myComputeMain", computeEntryPoint.writeRef());
+```
+
+### Composition
+
+An application might load any number of modules with `loadModule()`, and those modules might contain any number of entry points.
+Before GPU kernel code can be generated it is first necessary to decide which pieces of GPU code will be used together.
+
+Both `slang::IModule` and `slang::IEntryPoint` inherit from `slang::IComponentType`, because both can be used as components when composing a shader program.
+A composition can be created with `ISession::createCompositeComponentType()`:
+
+```c++
+IComponentType* components[] = { module, entryPoint };
+SlangComPtr<IComponentType> program;
+session->createCompositeComponentType(components, 2, program.writeRef());
+```
+
+As discussed earlier in this chapter, the composition operation serves two important purposes.
+First, it establishes which code is part of a compiled shader program and which is not.
+Second, it established an ordering for the code in a program, which can be used for layout.
+
+### Layout and Reflection
+
+Some applications need to perform reflection on shader parameters and their layout, whether at runtime or as part of an offline compilation tool.
+The Slang API allows layout to be queried on any `IComponentType` using `getLayout()`:
+
+```c++
+slang::ProgramLayout* layout = program->getLayout();
+```
+
+> #### Note ####
+> In  the current Slang API, the `ProgramLayout` type is not reference-counted.
+> Currently, the lifetime of a `ProgramLayout` is tied to the `IComponentType` that returned it.
+> An application must ensure that it retains the given `IComponentType` for as long as it uses the `ProgramLayout`.
+
+Note that because both `IModule` and `IEntryPoint` inherit from `IComponentType`, they can also be queried for their layouts individually.
+The layout for a module comprises just its global-scope parameters.
+The layout for an entry point comprises just its entry-point parameters (both `uniform` and varying).
+
+The details of how Slang computes layout, what guarantees it makes, and how to inspect the reflection information will be discussed in a later chapter.
+
+Because the layout computed for shader parameters may depend on the compilation target, the `getLayout()` method actually takes a `targetIndex` parameter that is the zero-based index of the target for which layout information is being queried.
```

Comment:
The targetIndex being either what was returned from adding a target, or the target specified within the createSessions index.

Reply:
This document doesn't show the variation of the API where targets are added one at a time, so only the latter case is applicable.

That said, I think that we should probably consider making two big changes to the COM-lite API for sessions:

1. We should probably get rid of the targets being part of the `SessionDesc` and instead have `ISession::addTarget()`, allowing people to add targets over the life of a session (there is no reason not to allow it, really)

2. The API entry points that currently take a `targetIndex` should probably just take an `ITarget*` instead, so that this kind of fiddly index-based stuff can be eliminated. We could preserve the behavior that a null `ITarget*`  is allowed for a session with only a single target (and make it an error otherwise).